### DISCUSSION TITLE
Tree Style Tabs slow to load on firefox startup with less than 10 tabs

### DISCUSSION THREAD

**Discussion by ghepzz** [piroor/treestyletab#3097] [Q&A]
hi, i've been using tree style tabs and firefox for years, until someday it slowed down
problem: takes around 5 seconds or so for less than 10 pinned tabs plus extras
before (i think 2019 or so): it was instant and firefox had more than 50 tabs, 

currently chrome also have 50 tabs, but also instantly loads the browser,

im pretty sure i haven't touched the settings, must be when firefox updated itself sometime ago 

i googled: tree style tabs too slow to load, and how to make firefox faster, 
i turned off "loading all tabs", 
refreshed firefox(a new copy) and then added tree style tab
and the problem is a little bit servere (although it was taking 10 or so seconds to load 10 pinned tabs) but now it takes 5 seconds to load more tabs which is more reasonable

running win10 64bit, 8gb ram thanks

**Comment by irvinm** [comment]
If you disable TST, does Firefox load as quickly as you expect with the same number of tabs?  (Thinking to see if this is more of a Firefox or a TST related item)

**Reply by neyder** [reply]
Hi, I got the very same problem, then disabled, tree style tab, and now Firefox tabs loads less tan a second after with 3 windows and many more tabs.

**Reply by Montagnard11** [reply]
Same here when restoring the windows from Firefox settings at startup or later from TabSessionManager plugin, even with the "Tree Style Tab" option OFF. Worth mentioning: I recently downgraded FF from v121 to 115ESR.
Restoring 9 windows, 264 tabs: forget it.

**Comment by neyder** [comment]
Hi, I got the very same problem, then disabled, tree style tab, and now firefox tabs loads less tan a second after with 3 windows and many more tabs.

**Comment by neyder** [comment]
I can confirm that delay on loading pages, also after disabling TST tabs loads as fast as they can.

**Comment by anhcoi123** [comment]
I got the same problem here. After I disabled Tree Style Tab, firefox run fast (normal) again. With TST on, it tooks ~2-3 Minutes for firefox to get ready on Firefox Startup. With TST off: about 10-20 seconds

**Comment by piroor** [comment]
Anyway I need more help with debugging experts... I could not reproduce such a painful slowing-down at the startup on my environment so I cannot investigate the root problem which triggers the slowing-down on my environment.

**Reply by notuo** [reply]
Hi. I am with the same issues (more than 20 minutes for less than 600 tabs). Disabling TST loads almost instantly

What can I do to help you? I am not expert but I can manage my way.

**Reply by piroor** [reply]
Thanks. Here is a document about debugging of TST itself: https://github.com/piroor/treestyletab/wiki/How-to-inspect-tree-of-tabs

There are two possible reasons of such too long delay on a tree restoration:

* It is because TST runs some infinite loop and it takes your CPU. [Collecting performance profile](https://github.com/piroor/treestyletab/wiki/How-to-inspect-tree-of-tabs#how-to-collect-a-performance-profile) looks the good first step to investigate this kind cases.
* It is because TST waits finishing of some asynchronous operations but it is not finished soon. Basically TST should restore tree of tabs after all tabs are restored by Firefox. Such an "waiting" can cause too long delay sometimes - such an asynchronous operations are generally implemented with Promise and it takes very large time to resolve a promises sometimes due to something unknown reasons. Detailed [debug logs](https://github.com/piroor/treestyletab/wiki/How-to-inspect-tree-of-tabs#how-to-collect-debug-logs-of-tst) may help investigate what causes such kind "unresolved Promise" problems.

**Reply by notuo** [reply]
Hi. Sent some files to you.
Regards

**Reply by piroor** [reply]
Thanks. Honestly, I meant that I hoped you fully joined to this tough debugging task and the file is inspected by yourself, anyway I've read the sent files.

* The performance profile graph is basically very low. It may mean that the slow loading is caused by (1) waiting for never resolved asynchronous operations, or (2) you failed to collect effective performance profile. When you start to collect the performance profile, did you choose "Firefox" (or "Nightly") from the "Settings" dropdown list below the "Start recording" button? If it was "Web Developer", the profiler cannot collect effective information on this case. Otherwise, if you collected the profile after the slow startup was finished, it was too late to start collecting performance profile.
* The logs from the console clearly indicate the startup task took too much time on your case. The final step of the initialization of the background page ([between two lines "1743: (async) init: initializing API for other addons" and "86892: (async) init: exporting tabs to sidebars"](https://github.com/piroor/treestyletab/blob/464fe788e90f1b380efdd80b3a8fb5903bbffd37/webextensions/background/background.js#L161-L172)) took 86 seconds (the number means he time from the previous step in milliseconds). And, in the sidebar page, the line ["243487: (async) importTabsFromBackground: kCOMMAND_PING_TO_BACKGROUND"](https://github.com/piroor/treestyletab/blob/464fe788e90f1b380efdd80b3a8fb5903bbffd37/webextensions/sidebar/sidebar.js#L630-L633) means that the background page took 243 seconds to [respond with tabs information](https://github.com/piroor/treestyletab/blob/464fe788e90f1b380efdd80b3a8fb5903bbffd37/webextensions/background/handle-misc.js#L563-L572).

From the investigation result, it looks that you need to collect new performance profile around the bottleneck operation, but it is not easy - we need to add some codes to wait in the middle of the initialization process of TST, until you started to collect performance profile, and you start to collect the performance profile while TST waits. Hmm...

**Reply by piroor** [reply]
OK, I've added the mechanism. Now you can collect performance profile around TST's initialization process with these steps:

1. Install Nightly (download the installer from https://www.mozilla.org/firefox/channel/desktop/ ) into somewhere different from your Firefox.
2. Copy your daily profile for Nightly.
   1. Start Nightly to prepare a new profile.
   2. Go to about:profiles.
   3. Find a profile with the message "This is the profile in use and it cannot be deleted." in the list.
   4. Click "Open Folder" of the "Root Directory" row. A folder window is opened, this is the profile folder of your Nightly.
   5. Start Firefox with your regular profile.
   6. Go to about:profiles.
   7. Find a profile with the message "This is the profile in use and it cannot be deleted." in the list.
   8. Click "Open Folder" of the "Root Directory" row. A folder window is opened, this is the profile folder of your regular Firefox.
   9. Exit both Nightly and Firefox.
   10. Delete all in the profile folder of your Nightly.
   11. Copy all files from the profile folder of your regular Firefox to the profile folder of your Nightly.
3. Start Nightly. TST's initialization may take too much time, but please wait until it finishes.
4. Install the development build of TST with the changes for debugging.
   1. Activate unsigned XPI installation. Go to about:config and set `xpinstall.signatures.required` to `false`.
   2. Download an unsigned XPI from https://github.com/piroor/treestyletab/actions/runs/13277026330 or more newer CI result.
   3. Install the unsigned XPI to your Nightly. TST's initialization may take too much time, but please wait until it finishes.
5. Activate blocking of TST's initialization.
   1. Go to about:debugging.
   2. Choose "This Nightly" form the left sidebar.
   3. Find "Tree Style Tab" under the section "Extensions".
   4. Click the "Inspect" button of the "Tree Style Tab" item.
   5. Switch to the "Console" tab.
   6. Run `configs.blockStartupOperations=true`. This activates the blocking feature of the initialization process.
6. Restart Nightly. You'll see TST sidebar with blank page.
7. Collect a performance profile around TST's initialization.
   1. Start to collect a performance profile with the setting "Nigthly".
   2. Go to about:debugging.
   3. Choose "This Nightly" form the left sidebar.
   4. Find "Tree Style Tab" under the section "Extensions".
   5. Click the "Inspect" button of the "Tree Style Tab" item.
   6. Switch to the "Console" tab.
   7. Run `configs.blockStartupOperations=false`. This triggers further initialization process of the TST.
   8. Wait until TST is completely initialized.
   9. Stop collecting of the performance profile.
